File: day1.py
def part1(input_string):
    return sum(int(digit) for i, digit in enumerate(input_string) if digit == input_string[(i + 1) % len(input_string)])


def part2(input_string):
    return sum(int(digit) for i, digit in enumerate(input_string) if digit == input_string[(i + len(input_string) / 2) % len(input_string)])


def main():
    with open('day1.txt') as f:
        input_string = f.read()

    print(part1(input_string))
    print(part2(input_string))

if __name__ == '__main__':
    main()

File: 2018/src/day1.rs
use std::collections::HashMap;
use std::fs::File;
use std::io::BufRead;
use std::io::BufReader;

fn part1() -> i32 {
  let f = File::open("files/day1.txt").unwrap();
  let file = BufReader::new(&f);

  let mut counter: i32 = 0;
  for (_, line) in file.lines().enumerate() {
    let num: i32 = line.unwrap().parse().expect("NaN");
    counter = counter + num;
  }

  return counter;
}

fn part2() -> i32 {
  let mut counter: i32 = 0;
  let mut values: HashMap<i32, bool> = HashMap::new();

  loop {
    let f = File::open("files/day1.txt").unwrap();
    let file = BufReader::new(&f);

    for (_, line) in file.lines().enumerate() {
      let num: i32 = line.unwrap().parse().expect("NaN");
      counter = counter + num;

      match values.get(&counter) {
        Some(_) => return counter,
        None => values.insert(counter, true),
      };
    }
  }
}

fn main() {
  println!("{}", part1());
  println!("{}", part2());
}
